include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -0,0 +1,146 @@
+
+    std::size_t __base_offset = std::max(__alignment, sizeof(__block_header));
+
+    // Check overflow on addition of __base_offset and __size
+    if (std::numeric_limits<std::size_t>::max() - __base_offset < __size) {
+        return nullptr;
+    }
+
+    // Memory block allocated with sycl::aligned_alloc_shared should be aligned to at least sizeof(__block_header) * 2
+    // to guarantee that header and header + sizeof(__block_header) (user pointer) would be placed in one memory page
+    std::size_t __usm_alignment = __base_offset << 1;
+    // Required number of bytes to store memory header and preserve alignment on returned pointer
+    // usm_alignment bytes are reserved to store memory header
+    std::size_t __usm_size = __size + __base_offset;
+
+    sycl::context __context = __device->get_platform().ext_oneapi_get_default_context();
+    void* __ptr = sycl::aligned_alloc_shared(__usm_alignment, __usm_size, *__device, __context);
+
+    if (__ptr != nullptr) {
+        void* __original_pointer = __ptr;
+        __ptr = static_cast<char*>(__ptr) + __base_offset;
+        __block_header* __header = static_cast<__block_header*>(__ptr) - 1;
+        *__header = __block_header{__uniq_type_const, __original_pointer, __device, __size};
+    }
+
+    return __ptr;
+}
+
+inline bool __same_memory_page(void* __ptr1, void* __ptr2) {
+    std::uintptr_t __page_size = __get_memory_page_size();
+    std::uintptr_t __page_mask = ~(__page_size - 1);
+    std::uintptr_t __ptr1_page_begin = std::uintptr_t(__ptr1) & __page_mask;
+    std::uintptr_t __ptr2_page_begin = std::uintptr_t(__ptr2) & __page_mask;
+    return __ptr1_page_begin == __ptr2_page_begin;
+}
+
+inline auto __get_original_realloc() {
+    using __realloc_func_type = void* (*)(void*, std::size_t);
+
+    static __realloc_func_type __orig_realloc = __realloc_func_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+
+inline void __free_usm_pointer(__block_header* __header) {
+    assert(__header != nullptr);
+    sycl::context __context = __header->_M_device->get_platform().ext_oneapi_get_default_context();
+    __header->_M_uniq_const = 0;
+    sycl::free(__header->_M_original_pointer, __context);
+}
+
+inline void* __realloc_real_pointer(void* __user_ptr, std::size_t __new_size) {
+    assert(__user_ptr != nullptr);
+    __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+
+    void* __result = nullptr;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const) {
+        if (__header->_M_requested_number_of_bytes >= __new_size) {
+            // No need to reallocate memory, previously allocated number of bytes is enough to store __new_size
+            __result = __user_ptr;
+        } else {
+            // Reallocate __new_size
+            assert(__header->_M_device != nullptr);
+            void* __new_ptr = __allocate_shared_for_device(__header->_M_device, __new_size, alignof(std::max_align_t));
```

Comment:
What is the expected behavior for realloc when the __user_ptr comes from outside of the translation unit? As written, it  will allocate the new memory using the device from the header. So if this ptr came from a translation unit A but winds up crossing over to translation unit B, it will still use the device from A when realloc is called from B. Is that what should happen? Or, should it use the device in B?

Similarly, a regular pointer from outside of the translation unit will not have a header. And so, if it is realloc-ed it will also not use the device from the translation unit that calls realloc. It would fall through to the else and re-use original realloc.  Is that the desired behavior?

I'm not sure what the behavior should be, just want to confirm that movement across translation units with different device policies selected has been considered in the current design.

Reply:
First of all, regarding passing memory between translation units with different devices - it is also a subject of additional limitations. For sure, it is prohibited to allocate memory in the TU with cpu device, then pass it into TU with gpu device and use it inside the body of some algorithm because memory allocated for cpu would not be available on gpu. Passing memory and using it only on the host side should not result in issues. 

Regarding `realloc` - it is intentional to preserve the memory type while doing reallocation. If the memory was allocated for cpu - it should be re-allocated for cpu also. Same for pointers came from regular malloc. My understanding of `realloc` usage is something like - "Oh, I just realized that the number of bytes I requested the last time is not enough, memory manager, please give me more bytes". Regular memory manager can also do it without releasing the memory. So we decided that changing the type of memory can be unexpected. 

And hence, allocating memory in the TU with cpu device, then passing it into TU with gpu device, calling `realloc` on it and then passing into the algorithm is also prohibited. I guess it should be mentioned somewhere in our documentation.